### ISSUE TITLE
Jump table identification does not resolve all jump targets

### ISSUE THREAD

**Issue by chq-matteo** [revng/revng#33]
# Reproduction steps

```
int switch_f(char c) {
    switch (c) {
        case 'A':
        case 'a':
            return 1;
        case 'B':
        case 'b':
            return 2;
        case 'C':
        case 'c':
            return 4;
        case 'D':
        case 'd':
            return 8;
        default:
            return 0;
    }
}
```

Compile using
```clang-10 -shared switch.c -o switch.c.debug.o```

Lift the binary with revng
```revng lift -g ll switch.c.debug.o.manual.ll```

Look into the lifted ir for `switch_f.0x30` and see that it jumps to the dispatcher

## Expectation

The jump table targets should be resolved

## Attachments

[switch.zip](https://github.com/revng/revng/files/5770893/switch.zip)

**Comment by aleclearmind** [comment]
I've a slightly different program:
```asm
0000000000000580 <switch_f>:
 580:   add    edi,0xffffffbf
 583:   cmp    edi,0x23
 586:   ja     5af <switch_f+0x2f>
 588:   mov    eax,0x1
 58d:   lea    rcx,[rip+0x2c]        # 5c0 <_fini+0xc>
 594:   movsxd rdx,DWORD PTR [rcx+rdi*4]
```

The problem is that we can prove that from 0x588 (and on) `edi <= 0x23`, but then `rdi` is used, for which we have no bounds.

```llvm
bb.0x400580:                                      ; preds = %dispatcher.entry_epoch_0_address_space_0_type_Code_x86_64
  %9 = add i64 %2, 4294967231
  %10 = trunc i64 %9 to i32
  %11 = icmp ugt i32 %10, 35
  br i1 %11, label %bb.0x4005af, label %bb.0x400588

bb.0x400588:
  %32 = shl i64 %9, 2
  %33 = and i64 %32, 17179869180
  %34 = add nuw nsw i64 %33, 4195776
  %35 = inttoptr i64 %34 to i32*
  call void (...) @revng_avi(i32* nonnull %35, i32 18), !revng.avi !86
```

Do you think type shrinking would be able to shrink to 32 bits `%33` and those defining it? `%9` in particular.
This would enable us to get a constraint on `%9` (since it would be collapsed with `%10`) and we should be good.

**Comment by chqma-revng** [comment]
If we use the intermediate results of the type shrinking we should know that only the lower `34` bits of `%33` contribute to the call argument and actually also that only the lower `32` bits of `%9` contribute to the call argument (unless they also flow into other sinks)

I think it would produce something like this before the clean up llvm passes that are scheduled after the proper `type-shrinking-pass`

```
bb.0x400580:                                      ; preds = %dispatcher.entry_epoch_0_address_space_0_type_Code_x86_64
  %ts0 = trunc i64 %2 to i32
  %9 = add i32 %ts0, 4294967231 ; <- we can prove that only 32 bits flow into sinks
  %ts1 = zext i64 %9
  %10 = trunc i64 %ts1 to i32 ; <- would probably be gone after clean up
  %11 = icmp ugt i32 %10, 35 ; <- would use %9 directly
  br i1 %11, label %bb.0x4005af, label %bb.0x400588

bb.0x400588:
  %32 = shl i64 %ts1, 2 ; <-- here we cannot down cast %32 to i32 because 34 bits are "alive" and same for the rest
  %33 = and i64 %32, 17179869180
  %34 = add nuw nsw i64 %33, 4195776
  %35 = inttoptr i64 %34 to i32*
  call void (...) @revng_avi(i32* nonnull %35, i32 18), !revng.avi !86
```

If you can give me the binary I can test this

**Comment by chqma-revng** [comment]
After the clean up, it would be something like this

```llvm
bb.0x400580:                                      ; preds = %dispatcher.entry_epoch_0_address_space_0_type_Code_x86_64
  %ts0 = trunc i64 %2 to i32
  %9 = add i32 %ts0, 4294967231 ; <- we can prove that only 32 bits flow into sinks
  %ts1 = zext i64 %9
  %11 = icmp ugt i32 %9, 35 ; <- would use %9 directly
  br i1 %11, label %bb.0x4005af, label %bb.0x400588

bb.0x400588:
  %32 = shl i64 %ts1, 2 ; <-- here we cannot down cast %32 to i32 because 34 bits are "alive" and same for the rest
  %33 = and i64 %32, 17179869180
  %34 = add nuw nsw i64 %33, 4195776
  %35 = inttoptr i64 %34 to i32*
  call void (...) @revng_avi(i32* nonnull %35, i32 18), !revng.avi !86
```

So if you can propagate the constraint from `%9` to its zero extended value then you should be good

**Comment by aleclearmind** [comment]
> ```llvm
> %ts1 = zext i64 %9
> ```

I would be *very* happy if this was in `bb.0x400588`.

**Comment by aleclearmind** [comment]
Fixed in 8cee47f!
Thanks!

**Comment by aleclearmind** [comment]
Unhandled situations are described in #52.
